File: arith.rb
# Modifying http://www.cis.upenn.edu/~bcpierce/tapl/checkers/arith/ for learning TAPL.
# TYPES AND PROGRAMMING LANGUAGES by Benjamin C. Pierce Copyright (c)2002 Benjamin C. Pierce

class Array
  def isnumericval
    t = self
    case
    when t[0] == :zero
      true
    when t[0] == :succ
      t[1].isnumericval
    else
      false
    end
  end

  def isval
    t = self
    case
    when t[0] == :true
      true
    when t[0] == :false
      true
    when t.isnumericval
      true
    else
      false
    end
  end

  class NoRuleApplies < Exception; end

  def eval1
    t = self
    case
    when t[0] == :if && t[1][0] == :true
      t[2]
    when t[0] == :if && t[1][0] == :false
      t[3]
    when t[0] == :if
      [:if,t[1].eval1,t[2],t[3]]
    when t[0] == :succ
      [:succ,t[1].eval1]
    when t[0] == :pred && t[1][0] == :zero
      [:zero]
    when t[0] == :pred && t[1][0] == :succ && t[1][1].isnumericval
      t[1][1]
    when t[0] == :pred
      [:pred,t[1].eval1]
    when t[0] == :iszero && t[1][0] == :zero
      [:true]
    when t[0] == :iszero && t[1][0] == :succ && t[1][1].isnumericval
      [:false]
    when t[0] == :iszero
      [:iszero,t[1].eval1]
    else
      raise NoRuleApplies
    end
  end

  def eval
    t = self
    begin
      t.eval1.eval
    rescue NoRuleApplies
      t
    end
  end
end

# Test
printf("test1: %s\n", [:true].eval == [:true])
printf("test2: %s\n", [:false].eval == [:false])
printf("test3: %s\n", [:if,[:true],[:true],[:false]].eval == [:true])
printf("test4: %s\n", [:if,[:false],[:true],[:false]].eval == [:false])
printf("test5: %s\n", [:if,[:if,[:true],[:true],[:false]],[:true],[:false]].eval == [:true])
printf("test6: %s\n", [:if,[:if,[:false],[:true],[:false]],[:true],[:false]].eval == [:false])
printf("test7: %s\n", [:zero].eval == [:zero])
printf("test8: %s\n", [:succ,[:pred,[:zero]]].eval == [:succ,[:zero]])
printf("test9: %s\n", [:iszero,[:succ,[:zero]]].eval == [:false])
printf("test10: %s\n", [:iszero,[:pred,[:succ,[:zero]]]].eval == [:true])
printf("test11: %s\n", [:iszero,[:pred,[:succ,[:succ,[:zero]]]]].eval == [:false])

File: arith.ml
(* Modifying http://www.cis.upenn.edu/~bcpierce/tapl/checkers/arith/ for learning TAPL. *)
(* TYPES AND PROGRAMMING LANGUAGES by Benjamin C. Pierce Copyright (c)2002 Benjamin C. Pierce *)

type term =
    TmTrue
  | TmFalse
  | TmIf of term * term * term
  | TmZero
  | TmSucc of term
  | TmPred of term
  | TmIsZero of term

let rec isnumericval t = match t with
    TmZero -> true
  | TmSucc(t1) -> isnumericval t1
  | _ -> false

let rec isval t = match t with
    TmTrue  -> true
  | TmFalse -> true
  | t when isnumericval t  -> true
  | _ -> false

exception NoRuleApplies

let rec eval1 t = match t with
    TmIf(TmTrue,t2,t3) ->
      t2
  | TmIf(TmFalse,t2,t3) ->
      t3
  | TmIf(t1,t2,t3) ->
      let t1' = eval1 t1 in
      TmIf(t1', t2, t3)
  | TmSucc(t1) ->
      let t1' = eval1 t1 in
      TmSucc(t1')
  | TmPred(TmZero) ->
      TmZero
  | TmPred(TmSucc(nv1)) when (isnumericval nv1) ->
      nv1
  | TmPred(t1) ->
      let t1' = eval1 t1 in
      TmPred(t1')
  | TmIsZero(TmZero) ->
      TmTrue
  | TmIsZero(TmSucc(nv1)) when (isnumericval nv1) ->
      TmFalse
  | TmIsZero(t1) ->
      let t1' = eval1 t1 in
      TmIsZero(t1')
  | _ -> 
      raise NoRuleApplies

let rec eval t =
  try let t' = eval1 t
      in eval t'
  with NoRuleApplies -> t

(* Test *)
let test1 = eval(TmTrue) = TmTrue;;
let test2 = eval(TmIf(TmFalse,TmTrue,TmFalse)) = TmFalse;;
let test3 = eval(TmZero) = TmZero;;
let test4 = eval(TmSucc(TmPred(TmZero))) = TmSucc(TmZero);;
let test5 = eval(TmIsZero(TmPred(TmSucc(TmSucc(TmZero))))) = TmFalse;;
